torch/csrc/distributed/rpc/ProcessGroupAgent.cpp

```diff
@@ -0,0 +1,227 @@
+  //    effort to fix this problem).
+  sync();
+  int dst = (pg_->getRank() + 1) % pg_->getSize();
+  enqueue(SendWork(dst, Message({}, {}, MessageType::SHUTDOWN)));
+  std::unique_lock<std::mutex> lock(sendQueueMutex_);
+  workConsumeCV_.wait(lock, [&] { return sendQueue_.empty(); });
+  stop_ = true;
+  lock.unlock();
+
+  workProduceCV_.notify_all();
+  sendThread_.join();
+  listenerThread_.join();
+}
+
+void ProcessGroupAgent::sync() {
+  // Block until all processes wants to sync. This is necessary before acquiring
+  // the lock below, because other processes might not enter sync() until it
+  // gets some response from this RpcAgent.
+  pg_->barrier()->wait();
+  // Acquire the lock on the send queue to prevent additional messages to be put
+  // onto the send queue.
+  std::unique_lock<std::mutex> lock(sendQueueMutex_);
+  // Wait until the send queue is depleted.
+  workConsumeCV_.wait(lock, [&] { return sendQueue_.empty(); });
+  // Use another barrier in case different RpcAgent handles different amounts of
+  // workloads.
+  pg_->barrier()->wait();
+}
+
+std::shared_ptr<FutureMessage> ProcessGroupAgent::send(
+    const std::string& to, Message&& message) {
+
+  auto dstRankIter = nameMap_.find(to);
+  TORCH_CHECK(dstRankIter != nameMap_.end(), "Unknown destination worker ", to);
+
+  const int dstRank = dstRankIter -> second;
+  TORCH_CHECK(dstRank != pg_->getRank(), "ProcessGroupAgent does not support "
+    "making RPC calls to self.")
+
+  auto requestId = nextId();
+  auto future = std::make_shared<FutureMessage>();
+  if (message.isRequest()) {
+    {
+      std::lock_guard<std::mutex> lock{futureMutex_};
+      futures_[requestId] = future;
+    }
+    message.setId(requestId);
+  } else {
+    future->markCompleted();
+  }
+
+  enqueue(SendWork(dstRank, std::move(message)));
+  return future;
+}
+
+void ProcessGroupAgent::enqueue(SendWork work) {
+  std::unique_lock<std::mutex> lock(sendQueueMutex_);
+  sendQueue_.emplace_back(std::move(work));
+  lock.unlock();
+
+  workProduceCV_.notify_one();
+}
+
+// making sure tensors are not deleted before send finishes
+void ProcessGroupAgent::sendLoop() {
+  std::unique_lock<std::mutex> lock(sendQueueMutex_);
+
+  while (!stop_) {
+    if (sendQueue_.empty()) {
+      workProduceCV_.wait(lock);
+      continue;
+    }
+
+    auto work = std::move(sendQueue_.front());
+    sendQueue_.pop_front();
+    lock.unlock();
+
+    workConsumeCV_.notify_one();
+
+
+    std::stringstream ss;
+    serialize(work.message_, ss);
+    std::string str = ss.str();
+
+    std::vector<torch::Tensor> preamble = {
+      torch::tensor(
+        {
+          (int64_t)pg_->getRank(),
+          (int64_t)str.length(),
+        }, {torch::kLong})
+    };
+    pg_->send(preamble, work.dstRank_, work.dstRank_ /* channelTag */)->wait();
+    std::vector<torch::Tensor> payload =
+        {torch::from_blob((void *)str.c_str(), str.length(), {torch::kChar})};
```

Comment:
Why do we serialize tensors into a string and then put them back into a tensor to send over the wire? Why don't we just create a bunch of tensors from the message (ex: payload, misc and the usual tensors) and just use pg_->send() on those tensors?